File: Minimal/tests/Brontide.Minimal.ComponentManagement.Tests/ActivationGroupTests.fs
namespace Brontide.Minimal.ComponentManagement.Tests

open System
open System.IO
open System.Text.Json
open NUnit.Framework
open Brontide.Minimal.Experimental.ComponentManagement

[<TestFixture>]
type ActivationGroupTests() =
    let peer = ContractId.create "brontide.fake.peer"
    let version = VersionLiteral.create "1.0"
    let first = OccurrenceId.create "occ.first"
    let second = OccurrenceId.create "occ.second"
    let region = RegionId.create "region.root"
    let multiple action = Assert.Multiple(Action action)

    let memberValue occurrence =
        { Occurrence = occurrence
          Definition = DefinitionId.create (sprintf "def.%s" (OccurrenceId.value occurrence))
          Region = region
          Provides = [ { Contract = peer; Version = version } ]
          RequiredReadyInputs = []
          AvailableReadyInputs = []
          WaitsForReadyOf = [] }

    let edge identity fromOccurrence toOccurrence =
        { Edge = ActivationEdgeId.create identity
          From = fromOccurrence
          To = toOccurrence
          Kind = ActivationDependencyKind.OrdinaryInteraction
          Contract = peer
          Version = version
          ObservedBeforeRelease = false
          Protocol = None
          CrossingPort = None
          AllowWiderRegionProposal = false }

    let emptyRequest members edges =
        { Request = ActivationGroupRequestId.create "activation.test"
          Generation = GenerationId.create "gen.test"
          RestartScope = RestartScopeId.create "restart.test"
          Members = members
          Edges = edges
          Protocols = []
          RegionCrossings = [] }

    let ordinaryCycle () =
        emptyRequest
            [ memberValue first; memberValue second ]
            [ edge "edge.first-second" first second
              edge "edge.second-first" second first ]

    let protocol (dependency: ActivationDependency) protocolId =
        { Protocol = protocolId
          Edge = dependency.Edge
          From = dependency.From
          To = dependency.To
          Operation =
            LifecycleOperationId.create
                (sprintf "operation.%s" (ActivationEdgeId.value dependency.Edge))
          Authority = [ CapabilityId.create "authority.lifecycle" ]
          InputShape = ShapeId.create "shape.lifecycle-input"
          OutputShape = ShapeId.create "shape.lifecycle-output"
          Ordering = "concurrent"
          TimeoutMilliseconds = 1000
          RetryLimit = 1
          Idempotent = true
          Completion = "peer-acknowledged"
          Failure = "fail-group"
          Rollback = "discard-provisional-state" }

    let relationalCycle () =
        let firstEdge =
            { edge "edge.first-second" first second with
                Kind = ActivationDependencyKind.RelationalInitialisation
                Protocol = Some(LifecycleProtocolId.create "protocol.first-second")
                ObservedBeforeRelease = true }
        let secondEdge =
            { edge "edge.second-first" second first with
                Kind = ActivationDependencyKind.RelationalInitialisation
                Protocol = Some(LifecycleProtocolId.create "protocol.second-first")
                ObservedBeforeRelease = true }
        { emptyRequest [ memberValue first; memberValue second ] [ firstEdge; secondEdge ] with
            Protocols =
                [ protocol firstEdge (Option.get firstEdge.Protocol)
                  protocol secondEdge (Option.get secondEdge.Protocol) ] }

    let rec permutations values =
        match values with
        | [] -> [ [] ]
        | _ ->
            [ for index in 0 .. List.length values - 1 do
                  let head = List.item index values
                  let tail =
                      values
                      |> List.indexed
                      |> List.choose (fun (candidateIndex, value) ->
                          if candidateIndex = index then None else Some value)
                  for suffix in permutations tail do
                      yield head :: suffix ]

    [<Test>]
    member _.``Neutral vector inventory is complete and data only``() =
        let path =
            Path.Combine(
                TestContext.CurrentContext.TestDirectory,
                "component-management",
                "fixtures",
                "cm3-activation-group-vectors.json")
        use document = JsonDocument.Parse(File.ReadAllText path)
        let root = document.RootElement
        let ids =
            root.GetProperty("vectors").EnumerateArray()
            |> Seq.map (fun vector -> vector.GetProperty("id").GetString())
            |> Seq.toArray
        multiple (fun () ->
            Assert.That(root.GetProperty("schemaVersion").GetInt32(), Is.EqualTo(1))
            Assert.That(root.GetProperty("fixture").GetString(), Is.EqualTo("cm3-activation-group-vectors"))
            Assert.That((ids = [| for index in 1 .. 18 -> sprintf "cm3-%02d" index |]), Is.True)
            Assert.That(root.GetRawText(), Does.Not.Contain("algorithm")))

    [<Test>]
    member _.``Ordinary cycle forms one Ready group without startup order or effects``() =
        match FakeActivationGroupPlanner.plan (ordinaryCycle ()) with
        | Planned plan ->
            let group = List.exactlyOne plan.Groups
            let expectedStages =
                [ ActivationStage.LocalInitialisation
                  ActivationStage.Interconnection
                  ActivationStage.Ready ]
            multiple (fun () ->
                Assert.That(group.Cyclic, Is.True)
                Assert.That(
                    ((group.Members |> List.map (fun item -> item.Occurrence)) = [ first; second ]),
                    Is.True)
                Assert.That(group.InternalEdges, Has.Length.EqualTo(2))
                Assert.That(group.Protocols, Is.Empty)
                Assert.That((group.Stages |> List.map (fun item -> item.Stage)) = expectedStages, Is.True)
                Assert.That(group.Stages |> List.forall (fun item -> not item.OrdinaryGateOpen), Is.True)
                Assert.That(group.ReleasePending, Is.True)
                Assert.That(plan.Effects, Is.EqualTo(Cm3EffectObservation.none)))
        | outcome -> Assert.Fail(sprintf "Expected activation-group plan, got %A" outcome)

    [<Test>]
    member _.``Complete bounded relational cycle adds only lifecycle stage``() =
        match FakeActivationGroupPlanner.plan (relationalCycle ()) with
        | Planned plan ->
            let group = List.exactlyOne plan.Groups
            let expectedStages =
                [ ActivationStage.LocalInitialisation
                  ActivationStage.Interconnection
                  ActivationStage.RelationalInitialisationStage
                  ActivationStage.Ready ]
            multiple (fun () ->
                Assert.That(group.Protocols, Has.Length.EqualTo(2))
                Assert.That((group.Stages |> List.map (fun item -> item.Stage)) = expectedStages, Is.True)
                Assert.That(group.Stages |> List.forall (fun item -> not item.OrdinaryGateOpen), Is.True)
                Assert.That(plan.Decisions |> List.filter (fun item -> item.Kind = "relational-protocol"), Has.Length.EqualTo(2)))
        | outcome -> Assert.Fail(sprintf "Expected relational plan, got %A" outcome)

    [<Test>]
    member _.``Descriptor cycle and version conflict fail without partial plan``() =
        let request = ordinaryCycle ()
        let descriptor =
            { request with
                Edges =
                    request.Edges
                    |> List.mapi (fun index item ->
                        if index = 0 then
                            { item with Kind = ActivationDependencyKind.DescriptorExpansion }
                        else item) }
            |> FakeActivationGroupPlanner.plan
        let conflict =
            { request with
                Edges =
                    request.Edges
                    |> List.mapi (fun index item ->
                        if index = 0 then
                            { item with Version = VersionLiteral.create "2.0" }
                        else item) }
            |> FakeActivationGroupPlanner.plan
        multiple (fun () ->
            match descriptor with
            | ActivationGroupRefused failure ->
                Assert.That(failure.Kind, Is.EqualTo(ActivationGroupFailureKind.RecursiveDescriptorExpansion))
            | outcome -> Assert.Fail(sprintf "Expected descriptor refusal, got %A" outcome)
            match conflict with
            | ActivationGroupRefused failure ->
                Assert.That(failure.Kind, Is.EqualTo(ActivationGroupFailureKind.ContractVersionConflict))
                Assert.That(failure.Source, Is.EqualTo(Some first))
                Assert.That(failure.Target, Is.EqualTo(Some second))
                Assert.That(failure.Contract, Is.EqualTo(Some peer))
                Assert.That(failure.Version, Is.EqualTo(Some(VersionLiteral.create "2.0")))
            | outcome -> Assert.Fail(sprintf "Expected version refusal, got %A" outcome)
            Assert.That(descriptor.Effects, Is.EqualTo(Cm3EffectObservation.none)))

    [<Test>]
    member _.``Lifecycle and ordinary gate violations are structured``() =
        let relational = relationalCycle ()
        let missing =
            { relational with
                Edges =
                    relational.Edges
                    |> List.mapi (fun index item ->
                        if index = 0 then { item with Protocol = None } else item) }
            |> FakeActivationGroupPlanner.plan
        let incomplete =
            { relational with
                Protocols =
                    relational.Protocols
                    |> List.mapi (fun index item ->
                        if index = 0 then { item with TimeoutMilliseconds = 0 } else item) }
            |> FakeActivationGroupPlanner.plan
        let ordinary = ordinaryCycle ()
        let early =
            { ordinary with
                Edges =
                    ordinary.Edges
                    |> List.mapi (fun index item ->
                        if index = 0 then { item with ObservedBeforeRelease = true } else item) }
            |> FakeActivationGroupPlanner.plan
        let undeclared =
            { ordinary with
                Edges =
                    ordinary.Edges
                    |> List.mapi (fun index item ->
                        if index = 0 then
                            { item with Protocol = Some(LifecycleProtocolId.create "protocol.unexpected") }
                        else item) }
            |> FakeActivationGroupPlanner.plan
        let oneRelationalEdge = relational.Edges.Head
        let crossGroup =
            { emptyRequest relational.Members [ oneRelationalEdge ] with
                Protocols = [ relational.Protocols.Head ] }
            |> FakeActivationGroupPlanner.plan
        let failureKind outcome =
            match outcome with
            | ActivationGroupRefused failure -> failure.Kind
            | value -> failwithf "Expected refusal, got %A" value
        multiple (fun () ->
            Assert.That(failureKind missing, Is.EqualTo(ActivationGroupFailureKind.LifecycleProtocolRequired))
            Assert.That(failureKind incomplete, Is.EqualTo(ActivationGroupFailureKind.LifecycleProtocolIncomplete))
            Assert.That(failureKind early, Is.EqualTo(ActivationGroupFailureKind.OrdinaryPreReleaseTraffic))
            Assert.That(failureKind undeclared, Is.EqualTo(ActivationGroupFailureKind.UndeclaredLifecycleTraffic))
            Assert.That(failureKind crossGroup, Is.EqualTo(ActivationGroupFailureKind.UndeclaredLifecycleTraffic)))

    [<Test>]
    member _.``Ready requires local inputs and acyclic wait graph``() =
        let request = ordinaryCycle ()
        let missing =
            { request with
                Members =
                    request.Members
                    |> List.mapi (fun index item ->
                        if index = 0 then
                            { item with RequiredReadyInputs = [ LifecycleInputId.create "input.missing" ] }
                        else item) }
            |> FakeActivationGroupPlanner.plan
        let waiting =
            { request with
                Members =
                    [ { request.Members[0] with WaitsForReadyOf = [ second ] }
                      { request.Members[1] with WaitsForReadyOf = [ first ] } ] }
            |> FakeActivationGroupPlanner.plan
        multiple (fun () ->
            match missing with
            | ActivationGroupRefused failure ->
                Assert.That(failure.Kind, Is.EqualTo(ActivationGroupFailureKind.ReadyInputUnavailable))
            | outcome -> Assert.Fail(sprintf "Expected input refusal, got %A" outcome)
            match waiting with
            | ActivationGroupRefused failure ->
                Assert.That(failure.Kind, Is.EqualTo(ActivationGroupFailureKind.CircularReadyWait))
                Assert.That(failure.Member, Is.Not.Null)
            | outcome -> Assert.Fail(sprintf "Expected Ready wait refusal, got %A" outcome))

    [<Test>]
    member _.``Cross Region cycle requires matching Port import export or widening``() =
        let port = PortId.create "port.peer"
        let childRegion = RegionId.create "region.child"
        let ordinary = ordinaryCycle ()
        let request =
            { ordinary with
                Members =
                    [ ordinary.Members[0]
                      { ordinary.Members[1] with Region = childRegion } ]
                Edges =
                    ordinary.Edges
                    |> List.map (fun item -> { item with CrossingPort = Some port }) }
        let crossings =
            request.Edges
            |> List.map (fun item ->
                { Edge = item.Edge
                  FromRegion = if item.From = first then region else childRegion
                  ToRegion = if item.To = second then childRegion else region
                  Port = port
                  ImportDeclared = true
                  ExportDeclared = true })
        let accepted =
            FakeActivationGroupPlanner.plan { request with RegionCrossings = crossings }
        let widened =
            FakeActivationGroupPlanner.plan
                { request with
                    Edges =
                        request.Edges
                        |> List.map (fun item -> { item with AllowWiderRegionProposal = true }) }
        let refused = FakeActivationGroupPlanner.plan request
        let conflict =
            FakeActivationGroupPlanner.plan
                { request with
                    RegionCrossings =
                        crossings
                        |> List.mapi (fun index item ->
                            if index = 0 then { item with ImportDeclared = false } else item) }
        multiple (fun () ->
            match accepted with
            | Planned plan ->
                Assert.That((List.exactlyOne plan.Groups).RegionCrossings, Has.Length.EqualTo(2))
                Assert.That(plan.RegionCrossings, Has.Length.EqualTo(2))
            | outcome -> Assert.Fail(sprintf "Expected cross-Region plan, got %A" outcome)
            match widened with
            | WiderParentGenerationRequired proposal -> Assert.That(proposal.Port, Is.EqualTo(port))
            | outcome -> Assert.Fail(sprintf "Expected widening proposal, got %A" outcome)
            match refused with
            | ActivationGroupRefused failure ->
                Assert.That(failure.Kind, Is.EqualTo(ActivationGroupFailureKind.RegionCrossingRequired))
            | outcome -> Assert.Fail(sprintf "Expected crossing refusal, got %A" outcome)
            match conflict with
            | ActivationGroupRefused failure ->
                Assert.That(failure.Kind, Is.EqualTo(ActivationGroupFailureKind.RegionCrossingConflict))
            | outcome -> Assert.Fail(sprintf "Expected crossing conflict, got %A" outcome))

    [<Test>]
    member _.``Acyclic condensation is dependency first and permutation invariant``() =
        let third = OccurrenceId.create "occ.third"
        let request =
            emptyRequest
                [ memberValue first; memberValue second; memberValue third ]
                [ edge "edge.first-second" first second
                  edge "edge.second-third" second third ]
        let baseline = FakeActivationGroupPlanner.plan request
        for members in permutations request.Members do
            for edges in permutations request.Edges do
                let outcome =
                    FakeActivationGroupPlanner.plan { request with Members = members; Edges = edges }
                Assert.That(outcome, Is.EqualTo(baseline))
        match baseline with
        | Planned plan ->
            let ordered =
                plan.Groups
                |> List.map (fun group -> (List.exactlyOne group.Members).Occurrence)
            Assert.That((ordered = [ third; second; first ]), Is.True)
        | outcome -> Assert.Fail(sprintf "Expected acyclic plan, got %A" outcome)

    [<Test>]
    member _.``Duplicate and missing identities fail closed``() =
        let request = ordinaryCycle ()
        let duplicate =
            FakeActivationGroupPlanner.plan
                { request with Members = request.Members @ [ request.Members.Head ] }
        let missing =
            FakeActivationGroupPlanner.plan
                { request with Members = [ request.Members.Head ] }
        let duplicateProvision =
            FakeActivationGroupPlanner.plan
                { request with
                    Members =
                        request.Members
                        |> List.mapi (fun index item ->
                            if index = 1 then
                                { item with Provides = item.Provides @ item.Provides }
                            else item) }
        let relational = relationalCycle ()
        let duplicateProtocolEdge =
            FakeActivationGroupPlanner.plan
                { relational with
                    Protocols =
                        relational.Protocols
                        @ [ { relational.Protocols.Head with
                                Protocol = LifecycleProtocolId.create "protocol.duplicate-edge" } ] }
        multiple (fun () ->
            match duplicate with
            | ActivationGroupRefused failure ->
                Assert.That(failure.Kind, Is.EqualTo(ActivationGroupFailureKind.ContradictoryIdentity))
            | outcome -> Assert.Fail(sprintf "Expected duplicate refusal, got %A" outcome)
            match missing with
            | ActivationGroupRefused failure ->
                Assert.That(failure.Kind, Is.EqualTo(ActivationGroupFailureKind.MissingMember))
            | outcome -> Assert.Fail(sprintf "Expected missing refusal, got %A" outcome)
            match duplicateProvision with
            | ActivationGroupRefused failure ->
                Assert.That(failure.Kind, Is.EqualTo(ActivationGroupFailureKind.ContractVersionConflict))
            | outcome -> Assert.Fail(sprintf "Expected duplicate provision refusal, got %A" outcome)
            match duplicateProtocolEdge with
            | ActivationGroupRefused failure ->
                Assert.That(failure.Kind, Is.EqualTo(ActivationGroupFailureKind.ContradictoryIdentity))
            | outcome -> Assert.Fail(sprintf "Expected duplicate protocol-edge refusal, got %A" outcome)
            Assert.That(missing.Effects, Is.EqualTo(Cm3EffectObservation.none)))

File: Reference/tests/Brontide.Reference.ComponentManagement.Tests/ActivationGroupTests.cs
using System.Text.Json;
using Brontide.Reference.Experimental.ComponentManagement;
using NUnit.Framework;

namespace Brontide.Reference.ComponentManagement.Tests;

[TestFixture]
public sealed class ActivationGroupTests
{
    private static readonly ContractId Peer = ContractId.Create("brontide.fake.peer");
    private static readonly VersionLiteral V1 = VersionLiteral.Create("1.0");
    private static readonly OccurrenceId First = OccurrenceId.Create("occ.first");
    private static readonly OccurrenceId Second = OccurrenceId.Create("occ.second");
    private static readonly RegionId Region = RegionId.Create("region.root");

    [Test]
    public void Neutral_vector_inventory_is_complete_and_data_only()
    {
        var path = Path.Combine(
            TestContext.CurrentContext.TestDirectory,
            "component-management",
            "fixtures",
            "cm3-activation-group-vectors.json");
        using var document = JsonDocument.Parse(File.ReadAllText(path));
        var root = document.RootElement;
        var ids = root.GetProperty("vectors")
            .EnumerateArray()
            .Select(vector => vector.GetProperty("id").GetString())
            .ToArray();

        Assert.Multiple(() =>
        {
            Assert.That(root.GetProperty("schemaVersion").GetInt32(), Is.EqualTo(1));
            Assert.That(root.GetProperty("fixture").GetString(), Is.EqualTo("cm3-activation-group-vectors"));
            Assert.That(ids, Is.EqualTo(Enumerable.Range(1, 18).Select(index => $"cm3-{index:00}").ToArray()));
            Assert.That(root.GetRawText(), Does.Not.Contain("algorithm"));
        });
    }

    [Test]
    public void Ordinary_cycle_forms_one_ready_group_without_startup_order_or_effects()
    {
        var outcome = new FakeActivationGroupPlanner().Plan(OrdinaryCycle());

        Assert.That(outcome.IsPlanned, Is.True);
        var group = outcome.Plan!.Groups.Single();
        Assert.Multiple(() =>
        {
            Assert.That(group.Cyclic, Is.True);
            Assert.That(group.Members.Select(item => item.Occurrence), Is.EqualTo(new[] { First, Second }));
            Assert.That(group.InternalEdges, Has.Count.EqualTo(2));
            Assert.That(group.Protocols, Is.Empty);
            Assert.That(group.Stages.Select(item => item.Stage), Is.EqualTo(new[]
            {
                ActivationStage.LocalInitialisation,
                ActivationStage.Interconnection,
                ActivationStage.Ready,
            }));
            Assert.That(group.Stages.All(item => !item.OrdinaryGateOpen), Is.True);
            Assert.That(group.ReleasePending, Is.True);
            Assert.That(outcome.Effects, Is.EqualTo(Cm3EffectObservation.None));
            Assert.That(outcome.Plan.Effects, Is.EqualTo(Cm3EffectObservation.None));
        });
    }

    [Test]
    public void Complete_bounded_relational_cycle_adds_only_the_lifecycle_stage()
    {
        var request = RelationalCycle();

        var outcome = new FakeActivationGroupPlanner().Plan(request);

        var group = outcome.Plan!.Groups.Single();
        Assert.Multiple(() =>
        {
            Assert.That(group.Protocols, Has.Count.EqualTo(2));
            Assert.That(group.Stages.Select(item => item.Stage), Is.EqualTo(new[]
            {
                ActivationStage.LocalInitialisation,
                ActivationStage.Interconnection,
                ActivationStage.RelationalInitialisation,
                ActivationStage.Ready,
            }));
            Assert.That(group.Stages.All(item => !item.OrdinaryGateOpen), Is.True);
            Assert.That(outcome.Plan.Decisions.Count(item => item.Kind == "relational-protocol"), Is.EqualTo(2));
        });
    }

    [Test]
    public void Descriptor_cycle_and_version_conflict_fail_without_partial_plan()
    {
        var cycle = OrdinaryCycle();
        var descriptor = new FakeActivationGroupPlanner().Plan(
            cycle with
            {
                Edges = cycle.Edges.Select((edge, index) =>
                    index == 0 ? edge with { Kind = ActivationDependencyKind.DescriptorExpansion } : edge).ToArray(),
            });
        var conflict = new FakeActivationGroupPlanner().Plan(
            cycle with
            {
                Edges = cycle.Edges.Select((edge, index) =>
                    index == 0 ? edge with { Version = VersionLiteral.Create("2.0") } : edge).ToArray(),
            });

        Assert.Multiple(() =>
        {
            Assert.That(descriptor.Failure!.Kind, Is.EqualTo(ActivationGroupFailureKind.RecursiveDescriptorExpansion));
            Assert.That(descriptor.Plan, Is.Null);
            Assert.That(conflict.Failure!.Kind, Is.EqualTo(ActivationGroupFailureKind.ContractVersionConflict));
            Assert.That(conflict.Failure.Source, Is.EqualTo(First));
            Assert.That(conflict.Failure.Target, Is.EqualTo(Second));
            Assert.That(conflict.Failure.Contract, Is.EqualTo(Peer));
            Assert.That(conflict.Failure.Version, Is.EqualTo(VersionLiteral.Create("2.0")));
            Assert.That(conflict.Plan, Is.Null);
            Assert.That(conflict.Effects, Is.EqualTo(Cm3EffectObservation.None));
        });
    }

    [Test]
    public void Lifecycle_and_ordinary_gate_violations_are_structured()
    {
        var relational = RelationalCycle();
        var missing = new FakeActivationGroupPlanner().Plan(
            relational with
            {
                Edges = relational.Edges.Select((edge, index) =>
                    index == 0 ? edge with { Protocol = null } : edge).ToArray(),
            });
        var incomplete = new FakeActivationGroupPlanner().Plan(
            relational with
            {
                Protocols = relational.Protocols.Select((protocol, index) =>
                    index == 0 ? protocol with { TimeoutMilliseconds = 0 } : protocol).ToArray(),
            });
        var ordinary = OrdinaryCycle();
        var early = new FakeActivationGroupPlanner().Plan(
            ordinary with
            {
                Edges = ordinary.Edges.Select((edge, index) =>
                    index == 0 ? edge with { ObservedBeforeRelease = true } : edge).ToArray(),
            });
        var undeclared = new FakeActivationGroupPlanner().Plan(
            ordinary with
            {
                Edges = ordinary.Edges.Select((edge, index) =>
                    index == 0 ? edge with { Protocol = LifecycleProtocolId.Create("protocol.unexpected") } : edge).ToArray(),
            });
        var oneRelationalEdge = relational.Edges[0];
        var crossGroup = new FakeActivationGroupPlanner().Plan(
            Empty(relational.Members, new[] { oneRelationalEdge }) with
            {
                Protocols = new[] { relational.Protocols[0] },
            });

        Assert.Multiple(() =>
        {
            Assert.That(missing.Failure!.Kind, Is.EqualTo(ActivationGroupFailureKind.LifecycleProtocolRequired));
            Assert.That(incomplete.Failure!.Kind, Is.EqualTo(ActivationGroupFailureKind.LifecycleProtocolIncomplete));
            Assert.That(early.Failure!.Kind, Is.EqualTo(ActivationGroupFailureKind.OrdinaryPreReleaseTraffic));
            Assert.That(undeclared.Failure!.Kind, Is.EqualTo(ActivationGroupFailureKind.UndeclaredLifecycleTraffic));
            Assert.That(crossGroup.Failure!.Kind, Is.EqualTo(ActivationGroupFailureKind.UndeclaredLifecycleTraffic));
        });
    }

    [Test]
    public void Ready_requires_local_inputs_and_an_acyclic_wait_graph()
    {
        var request = OrdinaryCycle();
        var missing = new FakeActivationGroupPlanner().Plan(
            request with
            {
                Members = request.Members.Select((member, index) =>
                    index == 0
                        ? member with { RequiredReadyInputs = new[] { LifecycleInputId.Create("input.missing") } }
                        : member).ToArray(),
            });
        var waiting = new FakeActivationGroupPlanner().Plan(
            request with
            {
                Members = new[]
                {
                    request.Members[0] with { WaitsForReadyOf = new[] { Second } },
                    request.Members[1] with { WaitsForReadyOf = new[] { First } },
                },
            });

        Assert.Multiple(() =>
        {
            Assert.That(missing.Failure!.Kind, Is.EqualTo(ActivationGroupFailureKind.ReadyInputUnavailable));
            Assert.That(waiting.Failure!.Kind, Is.EqualTo(ActivationGroupFailureKind.CircularReadyWait));
            Assert.That(waiting.Failure.Member, Is.Not.Null);
            Assert.That(waiting.Plan, Is.Null);
        });
    }

    [Test]
    public void Cross_region_cycle_requires_matching_import_export_or_explicit_widening()
    {
        var port = PortId.Create("port.peer");
        var request = OrdinaryCycle();
        request = request with
        {
            Members = new[]
            {
                request.Members[0],
                request.Members[1] with { Region = RegionId.Create("region.child") },
            },
            Edges = request.Edges.Select(edge => edge with { CrossingPort = port }).ToArray(),
        };
        var crossings = request.Edges.Select(edge => new RegionCrossingDeclaration(
            edge.Edge,
            edge.From == First ? Region : RegionId.Create("region.child"),
            edge.To == Second ? RegionId.Create("region.child") : Region,
            port,
            true,
            true)).ToArray();
        var accepted = new FakeActivationGroupPlanner().Plan(request with { RegionCrossings = crossings });
        var widened = new FakeActivationGroupPlanner().Plan(
            request with
            {
                Edges = request.Edges.Select(edge => edge with { AllowWiderRegionProposal = true }).ToArray(),
            });
        var refused = new FakeActivationGroupPlanner().Plan(request);
        var conflict = new FakeActivationGroupPlanner().Plan(
            request with
            {
                RegionCrossings = crossings.Select((crossing, index) =>
                    index == 0 ? crossing with { ImportDeclared = false } : crossing).ToArray(),
            });

        Assert.Multiple(() =>
        {
            Assert.That(accepted.IsPlanned, Is.True);
            Assert.That(accepted.Plan!.Groups.Single().RegionCrossings, Has.Count.EqualTo(2));
            Assert.That(accepted.Plan.RegionCrossings, Has.Count.EqualTo(2));
            Assert.That(
                () => ((IList<RegionCrossingDeclaration>)accepted.Plan.RegionCrossings)[0] = crossings[0],
                Throws.TypeOf<NotSupportedException>());
            Assert.That(widened.Wider!.Port, Is.EqualTo(port));
            Assert.That(widened.Plan, Is.Null);
            Assert.That(refused.Failure!.Kind, Is.EqualTo(ActivationGroupFailureKind.RegionCrossingRequired));
            Assert.That(conflict.Failure!.Kind, Is.EqualTo(ActivationGroupFailureKind.RegionCrossingConflict));
        });
    }

    [Test]
    public void Acyclic_condensation_is_dependency_first_and_permutation_invariant()
    {
        var third = OccurrenceId.Create("occ.third");
        var request = Empty(
            new[] { Member(First), Member(Second), Member(third) },
            new[]
            {
                Edge("edge.first-second", First, Second),
                Edge("edge.second-third", Second, third),
            });
        var baseline = Canonical(new FakeActivationGroupPlanner().Plan(request));

        foreach (var members in Permutations(request.Members))
        {
            foreach (var edges in Permutations(request.Edges))
            {
                var outcome = new FakeActivationGroupPlanner().Plan(request with { Members = members, Edges = edges });
                Assert.That(Canonical(outcome), Is.EqualTo(baseline));
            }
        }

        var groups = new FakeActivationGroupPlanner().Plan(request).Plan!.Groups;
        Assert.That(
            groups.Select(group => group.Members.Single().Occurrence),
            Is.EqualTo(new[] { third, Second, First }));
    }

    [Test]
    public void Duplicate_and_missing_identities_fail_closed()
    {
        var request = OrdinaryCycle();
        var duplicate = new FakeActivationGroupPlanner().Plan(
            request with { Members = request.Members.Concat(request.Members.Take(1)).ToArray() });
        var missing = new FakeActivationGroupPlanner().Plan(
            request with { Members = request.Members.Take(1).ToArray() });
        var duplicateProvision = new FakeActivationGroupPlanner().Plan(
            request with
            {
                Members = request.Members.Select((member, index) =>
                    index == 1
                        ? member with { Provides = member.Provides.Concat(member.Provides).ToArray() }
                        : member).ToArray(),
            });
        var relational = RelationalCycle();
        var duplicateProtocolEdge = new FakeActivationGroupPlanner().Plan(
            relational with
            {
                Protocols = relational.Protocols.Concat(new[]
                {
                    relational.Protocols[0] with
                    {
                        Protocol = LifecycleProtocolId.Create("protocol.duplicate-edge"),
                    },
                }).ToArray(),
            });

        Assert.Multiple(() =>
        {
            Assert.That(duplicate.Failure!.Kind, Is.EqualTo(ActivationGroupFailureKind.ContradictoryIdentity));
            Assert.That(missing.Failure!.Kind, Is.EqualTo(ActivationGroupFailureKind.MissingMember));
            Assert.That(duplicateProvision.Failure!.Kind, Is.EqualTo(ActivationGroupFailureKind.ContractVersionConflict));
            Assert.That(duplicateProtocolEdge.Failure!.Kind, Is.EqualTo(ActivationGroupFailureKind.ContradictoryIdentity));
            Assert.That(duplicate.Effects, Is.EqualTo(Cm3EffectObservation.None));
            Assert.That(missing.Plan, Is.Null);
        });
    }

    [Test]
    public void Returned_nested_collections_are_read_only_and_detached()
    {
        var provides = new[] { new ProvidedContract(Peer, V1) };
        var first = Member(First) with { Provides = provides };
        var request = Empty(new[] { first, Member(Second) }, new[] { Edge("edge.first-second", First, Second) });

        var outcome = new FakeActivationGroupPlanner().Plan(request);
        provides[0] = new ProvidedContract(ContractId.Create("brontide.fake.changed"), V1);

        var returned = outcome.Plan!.Groups.Single(group => group.Members.Any(item => item.Occurrence == First))
            .Members.Single(item => item.Occurrence == First);
        Assert.Multiple(() =>
        {
            Assert.That(returned.Provides.Single().Contract, Is.EqualTo(Peer));
            Assert.That(
                () => ((IList<ProvidedContract>)returned.Provides)[0] = new ProvidedContract(Peer, V1),
                Throws.TypeOf<NotSupportedException>());
        });
    }

    private static ActivationGroupRequest OrdinaryCycle() =>
        Empty(
            new[] { Member(First), Member(Second) },
            new[]
            {
                Edge("edge.first-second", First, Second),
                Edge("edge.second-first", Second, First),
            });

    private static ActivationGroupRequest RelationalCycle()
    {
        var edges = new[]
        {
            Edge("edge.first-second", First, Second) with
            {
                Kind = ActivationDependencyKind.RelationalInitialisation,
                Protocol = LifecycleProtocolId.Create("protocol.first-second"),
                ObservedBeforeRelease = true,
            },
            Edge("edge.second-first", Second, First) with
            {
                Kind = ActivationDependencyKind.RelationalInitialisation,
                Protocol = LifecycleProtocolId.Create("protocol.second-first"),
                ObservedBeforeRelease = true,
            },
        };
        return Empty(
            new[] { Member(First), Member(Second) },
            edges) with
        {
            Protocols = edges.Select(edge => Protocol(edge, edge.Protocol!.Value)).ToArray(),
        };
    }

    private static ActivationGroupRequest Empty(
        IReadOnlyList<ActivationGroupMember> members,
        IReadOnlyList<ActivationDependency> edges) =>
        new(
            ActivationGroupRequestId.Create("activation.test"),
            GenerationId.Create("gen.test"),
            RestartScopeId.Create("restart.test"),
            members,
            edges,
            Array.Empty<LifecycleProtocolDeclaration>(),
            Array.Empty<RegionCrossingDeclaration>());

    private static ActivationGroupMember Member(OccurrenceId occurrence) =>
        new(
            occurrence,
            DefinitionId.Create($"def.{occurrence.Value}"),
            Region,
            new[] { new ProvidedContract(Peer, V1) },
            Array.Empty<LifecycleInputId>(),
            Array.Empty<LifecycleInputId>(),
            Array.Empty<OccurrenceId>());

    private static ActivationDependency Edge(string id, OccurrenceId from, OccurrenceId to) =>
        new(
            ActivationEdgeId.Create(id),
            from,
            to,
            ActivationDependencyKind.OrdinaryInteraction,
            Peer,
            V1,
            false,
            null,
            null,
            false);

    private static LifecycleProtocolDeclaration Protocol(
        ActivationDependency edge,
        LifecycleProtocolId protocol) =>
        new(
            protocol,
            edge.Edge,
            edge.From,
            edge.To,
            LifecycleOperationId.Create($"operation.{edge.Edge.Value}"),
            new[] { CapabilityId.Create("authority.lifecycle") },
            ShapeId.Create("shape.lifecycle-input"),
            ShapeId.Create("shape.lifecycle-output"),
            "concurrent",
            1000,
            1,
            true,
            "peer-acknowledged",
            "fail-group",
            "discard-provisional-state");

    private static string Canonical(ActivationGroupOutcome outcome) =>
        JsonSerializer.Serialize(outcome);

    private static IEnumerable<T[]> Permutations<T>(IReadOnlyList<T> values)
    {
        if (values.Count == 0)
        {
            yield return Array.Empty<T>();
            yield break;
        }

        for (var index = 0; index < values.Count; index++)
        {
            var head = values[index];
            var tail = values.Where((_, candidateIndex) => candidateIndex != index).ToArray();
            foreach (var suffix in Permutations(tail))
            {
                yield return new[] { head }.Concat(suffix).ToArray();
            }
        }
    }
}
